EIPS/eip-1066.md

```diff
@@ -0,0 +1,314 @@
++---------------------------+           |                         |                         |                   |
+|                           |           |                         |                         |                   |
+| Updates ID with provider  |           |                         |                         |                   |
+|                           |           |                         |                         |                   |
++---------------------------+           |                         |                         |                   |
+             |                          |                         |                         |                   |
+             |         buy()            |                         |                         |                   |
+             +------------------------> |        check()          |                         |                   |
+             |                          +-----------------------> |         check()         |                   |
+             |                          |                         +-----------------------> |                   |
+             |                          |                         |                         |                   |
+             |                          |                         |       Status [0x11]     |                   |
+             |                          |                         | <-----------------------+                   |
+             |                          |                         |                         |                   |
+             |                          |                         |                         |   check()         |
+             |                          |                         +-------------------------------------------> |
+             |                          |                         |                         |                   |
+             |                          |                         |                         |  Status [0x11]    |
+             |                          |       Status [0x11]     | <-------------------------------------------+
+             |        Status [0x01]     | <-----------------------+                         |                   |
+             | <------------------------+                         |                         |                   |
+             |                          |                         |                         |                   |
+             |                          |                         |                         |                   |
+             |                          |                         |                         |                   |
+             +                          +                         +                         +                   +
+```
+
+## Rationale
+
+### Encoding
+
+ESCs are encoded as a `byte`. Hex values break nicely into high and low nibbles:
```

Comment:
Given existing ABI encoding expands all values to word size (256 bits / 32 bytes), it might make sense to use a larger type for this, providing more flexibility.

Reply:
Hm, yes perhaps for transmission format. Will change 👍 

However, we feel that the actual codes themselves should be limited to one `byte`. 256 codes is _plenty_, and showing restraint gives us a number of advantages:

* 256 codes is already _plenty_ to express many ideas, and to handle as return values
* Easy to decode 16x16 without a reference guide once you're familiar with them
* 32-bits would be ~4.3 billion codes, which largely defeats the purpose of a finite set
* You can pack multiple single `byte`s into a `bytes32` for cases where multiple codes are appropriate (though what to do about padding is another question)

Reply:
Using a wider type would permit application-specific return codes without collisions, which seems useful to me.

I don't think people should be returning multiple status codes; that seems liable to complicate matters for little or no gain.

Reply:
Sorry for the late reply; it's been a bit of a crazy week. I really do appreciate you hashing out these details 🙏

> I don't think people should be returning multiple status codes; that seems liable to complicate matters for little or no gain.

👍 I actually agree about the multiple status code. This is something that keeps coming up as an ask in local community conversations, but I'm not a fan. It's good to hear that I'm not the only one!

> Using a wider type would permit application-specific return codes without collisions, which seems useful to me.

I'm still not sure if such a large space for application-specific codes would be useful or harmful. Perhaps if you could provide some use cases for such a large space. Below is some more detailed reasoning on why I believe that 2^8 codes is the right call.

I'm not totally convinced about the added space, especially since (if?) people will be choosing codes, not a random hash. A random hash would be inconsistent with a semantic code system, which in turn adds friction. People will tend towards things that are easy to remember, or have some other value. For example: `0xCAFE` and `0xBAD_` may get a lot of use.

Even with a random hash for say, the first two digits of four, I question why they need to live in the same namespace. A contract that forwards other codes really has that as part of it's contract (in the [DbC](https://en.wikipedia.org/wiki/Design_by_contract) sense), and needs to maintain them itself. Law of Demeter aside, an explosion of forwarded codes makes it very difficult to know what to expect, and thus limits the usefulness of such a system.

I totally get that `bytes32` is one machine word, and padding a `byte` with a bunch of `0`s to fill out the word can seem like a waste of space. The purpose of this design is to be constrained for _developers_, with the goal of making it easy to choose a code, for other contracts to be able to react autonomously, and to translate messages. These all require a manageable number of responses to be effective. With `bytes32`, we're allocating a very small fraction of a percent to known codes, and the other [75 nines] would be unspecified. With 256 status known codes, if someone needs to use another scheme for their application, they're still free to. Adding 2^256 (2^32*8) codes to the same space seems to defeat the purpose. 

If we later wanted to expand the code-space, `bytes1` casts to `bytes32`, so we have simple interoperability between the ESCv0 and the hypothetical ESCv1.

Phew, that ended up being a bit longer than intended 😅 Not trying to brow beat with walls of text; just trying to clarify the intent and design!